File: Baekjun/24.8.1/1201.NMK/solution.py
from collections import deque

N, M, K = map(int, input().split())

if N > M * K or N - K < M - 1:
    print(-1)
    exit(0)

remaining_elements = deque(range(K + 1, N + 1))
result = list(range(K, 0, -1))

if M - 1 == 0:
    print(" ".join(map(str, result)))
    exit(0)

group_size, extra_elements = divmod(N - K, M - 1)

while remaining_elements:
    current_group = []
    for _ in range(group_size):
        current_group.append(remaining_elements.popleft())
    if extra_elements > 0:
        current_group.append(remaining_elements.popleft())
        extra_elements -= 1
    result.extend(current_group[::-1])

print(" ".join(map(str, result)))

File: Baekjun/24.8.1/1201.NMK/solution.rs
use std::collections::VecDeque;
use std::io::{self};

fn main() {
    // 표준 입력을 처리
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let parts: Vec<usize> = input
        .trim()
        .split_whitespace()
        .map(|x| x.parse().unwrap())
        .collect();
    let (n, m, k) = (parts[0], parts[1], parts[2]);

    // 조건 검사
    if n > m * k || n - k < m - 1 {
        println!("-1");
        return;
    }

    // 초기 배열 생성
    let mut remaining_elements: VecDeque<usize> = ((k + 1)..=n).collect();
    let mut result: Vec<usize> = (1..=k).rev().collect();

    // 전체가 하나의 그룹일 경우 바로 출력
    if m - 1 == 0 {
        println!(
            "{}",
            result
                .iter()
                .map(|x| x.to_string())
                .collect::<Vec<String>>()
                .join(" ")
        );
        return;
    }

    // 그룹 분할 계산
    let (group_size, mut extra_elements) = ((n - k) / (m - 1), (n - k) % (m - 1));

    // 그룹별 요소 추가
    while !remaining_elements.is_empty() {
        let mut current_group = Vec::new();
        for _ in 0..group_size {
            if let Some(val) = remaining_elements.pop_front() {
                current_group.push(val);
            }
        }
        if extra_elements > 0 {
            if let Some(val) = remaining_elements.pop_front() {
                current_group.push(val);
            }
            extra_elements -= 1;
        }
        current_group.reverse();
        result.extend(current_group);
    }

    // 결과 출력
    println!(
        "{}",
        result
            .iter()
            .map(|x| x.to_string())
            .collect::<Vec<String>>()
            .join(" ")
    );
}
